### PR TITLE
feat(telemetry): record the retry prompt and tool arguments in Level 3 content

### PR DESCRIPTION
**Pull request by dhshah13**
Third and last PR of the ADR 0050 Level 3 series, after #6429 (gate, collector, budget) and #6603 (tool results, `execute_tool` spans, ADR 0108). It delivers the two items #6603's description named as next: the runner-composed retry prompt as `gen_ai.input.messages`, and tool-call `arguments` on the message record. The carrier is unchanged — the record on the `agent` span, as ADR 0108 decided; `execute_tool` spans stay metadata only. Still one env var, no new configuration.

## What this does

| Part | Change |
|---|---|
| Input (`internal/cli`) | On a retry iteration under `validation_loop.feedback_mode: append`, the prompt `buildFeedbackPrompt` composed is recorded on that iteration's `agent` span as `gen_ai.input.messages`: one `user` message, one text part (semconv v1.37.0 input-messages schema). It is set right after the span starts, so the cancel, error and normal finalize paths carry it. The first iteration and retries without feedback send the runtime's fixed default prompt and record nothing. |
| Parser (`internal/runtime`) | `ToolUseEvent` gains `Arguments`: the call's input as the stream carried it, from both Claude Code paths (assistant line; `stream_event` deltas). Unchecked and unredacted at this layer, never rendered. Input assembled from deltas is cut at the parser's 1 MiB `maxToolInputSize`. pi and codex leave it empty (#7414); so does OpenCode, a stub runtime. |
| Collector (`internal/cli`) | `tool_call` parts carry `arguments`. The value is decoded, each string and object key is redacted on its own, each string member is scanned once more — already redacted — beside its key (only the member-name pattern's finding is kept from that scan), and the result is encoded again. Done once, when the event is handled. A call without a name — or whose name redacts to nothing — carries none; arguments lost to a name that redacted away are charged and marked like the dropped ones below. |
| Bounds | New `maxToolArgumentsBytes = 8 KiB`, measured on the redacted encoding. Over it — or when the text is not one JSON value, or two keys of one object redact to the same string — the arguments are dropped whole (a cut object is not JSON): the part keeps id, name and summary, is marked `fullsend.truncated`, and the bytes are charged to `fullsend.content.dropped_bytes`. The input message's encoded size is charged against the existing 255,000-byte ceiling, so the two content attributes of one span stay within the proven size together. |
| Redaction | On both sides of the pattern pipeline, the collector replaces the value of each sensitive runner env key (8 bytes or longer) with `[REDACTED:<key>]` — `replaceEnvSecrets`, the literal pass `redactFeedback` had, now shared by both. It applies to all Level 3 content, the text and tool results #6429 and #6603 shipped included. A call loses its `summary` when redaction found a secret in its arguments; a number leaf in arguments is scanned as its digits. |
| Docs | Reference, dev guide, runtime matrix, two user guides, the eval-measurements Level 3 caveat; dated annotations on ADR 0050 and ADR 0108 (no Decision text changed). |

## Why arguments are not scanned as serialized text

The redactor's patterns are written for plain text. A table test run first against the simple approach (scan the JSON text, keep it if still valid) failed 4 of the 8 shapes it started with (the first eight rows of `TestContentCollector_RedactsSecretsInArguments`; the table has since grown): an assignment that opens a string (`{"command":"DEPLOY_TOKEN=… make"}` — the pattern anchors on line start or whitespace, and sees a quote), an assignment after an escaped newline, a value behind escaped quotes, and JSON nested in a string. Unicode folding over the text also turned fullwidth quotation marks into ones that closed the string and added a member. Decoding first gives the redactor the text it was written for; the per-member scan keeps what the serialized form did catch (`"password": "…"`), where the leaf alone has no context. That scan runs on every string member, not only clean ones — a zero-width character, an accent or a second token in the value would otherwise switch it off — it goes to the pattern stage alone (both strings are already normalized, and a second normalizer pass over the pair can strip an escape sequence the first left open, and the value with it), and its probe text puts a comma where a double quote was: the quote would end the pattern's quoted run early, and the comma — like the quote — is outside every pattern's token class, so it joins no two runs into a token a prefix pattern would mask first (an underscore would: it sits inside most token classes).

## Measured

Three captured live review streams (the ones behind #6603's figures), sub-agent calls included:

| run | calls | arguments total | p50 | p90 | max | over 8 KiB |
|---|---|---|---|---|---|---|
| 32869162122 | 255 | 81 KB | 110 B | 211 B | 12.4 KB | 1 |
| 32871702429 | 141 | 100 KB | 108 B | 279 B | 24.5 KB | 3 |
| 32873411835 | 117 | 41 KB | 109 B | 280 B | 8.0 KB | 0 |

Every call over 8 KiB is an `Agent` dispatch prompt. Replayed through the real collector: with arguments sharing the 256 KiB total, the shipped bounds evict 31–63% of tool results (28–56% before — the baseline replay gives #6603's published 55% / 56% / 28%, in this table's run order); a 1 MiB total still evicts none (records of 456–1,011 KB); encoding adds 8–10% at these bounds. The per-member scan added no finding on the 513 real calls (53 / 40 / 16 findings per run with and without arguments). **Not measured:** an agent that writes files — a `Write` call carries the file body as an argument. Raising the total stays #7415.

## Decisions

- **Dropped whole, not cut.** `tool_call` parts keep the property #6603 gave them — never cut, only dropped — so the exact-accounting tests stand; the property test now generates arguments and counts them. The cost on the three measured review runs: the 0–3 `Agent` prompts per iteration over 8 KiB are not recorded. On an agent that writes files — not measured; in this repo `fix` runs on Claude Code, and `review` now runs on pi, which records no arguments until #7414 — it is any `Write` or `Edit` whose arguments re-encode over 8 KiB: the part keeps its `summary` (the file path) and nothing of the body, which is the case #6603 gave as the reason for the bound. Open to reviewer input, like the tail-vs-head question on results.
- **The recorded input is the redacted copy, not the bytes the agent received.** `redactFeedback` scans the feedback before it is sanitized and framed, and sanitizing can join a token that scan saw split, so the recorded copy is pattern-scanned again (the prompt the agent is sent is not changed); the pass also folds compatibility characters the prompt kept. Folding can grow the copy about elevenfold (U+FDFA, 3 bytes to 33): worst case about 113 KB for a 10 KiB feedback (113,095 bytes measured), which is why the size is charged to the ceiling rather than assumed small; a test builds that case and pins that input and output together stay within 255,000 bytes.
- **`summary` stays** beside `arguments`: it is what survives when arguments are dropped, and what pi, codex and OpenCode have.
- **Findings are one count.** Input, arguments and output findings all land in `fullsend.content.redactions`; `Result` no longer returns early when an iteration has findings or dropped bytes and no parts (a retry that fails before emitting anything). The input's pattern scan repeats `redactFeedback`'s when sanitizing changed nothing; a connection-string mask left by that first scan matches its own pattern and counts a finding. Secrets `redactFeedback` already masked are recorded as masks and are not counted (it discards its findings), so a retry prompt carrying `ghp_...` adds 0 to `fullsend.content.redactions`; `TestAttachInput_RedactsAndCountsFindings` feeds `attachInput` an unmasked token, the case where this scan is the first to see it.
- **Keys that redact alike drop the arguments.** Folding can make two keys of one object the same string; keeping either member would show a call the agent did not make, so the arguments go the way of over-bound ones. Keys are walked in sorted order so neither findings nor the charge follow map order.
- **Tool results stay on the output record.** `gen_ai.input.messages` carries one message, the runner's `user` prompt. The convention's worked example puts client-executed tool results in that attribute under `role:"tool"`; here they stay on `gen_ai.output.messages`, so the role-placement deviation #6603 documented at `contentMessage` is unchanged, for the reasons given there and in ADR 0108: parts keep stream order, and that record is the scorer contract.
- **Runner env values are replaced, not only patterns matched.** The sandbox is not meant to hold `runner_env` credentials, but the record leaves the runner, and the contribution guide's first redaction invariant asks for the literal pass wherever content does. The pass runs ahead of the pipeline, on the text as written, so a pattern does not mask part of a value and keep its first bytes, and again on the pipeline's result, because its normalizer joins a value the stream split with an invisible character or spelled in compatibility forms. Not covered, and pinned by a test for the assignment case: a value written that way which a pattern also recognises — in an assignment, a header, a secret-named field or a connection string, or by its own prefix — is masked by that pattern, and shows what its mask shows. Longer values go first, so a value that contains another is replaced whole and the text does not depend on map order (this changes `redactFeedback` too). Each replaced key counts once per pass per scanned string in `fullsend.content.redactions`, and again where a pattern then masks the marker (an assignment, a header, a secret-named field); an id that holds a value is dropped, like an id with any other finding. The parser cuts a call's `summary` out of its arguments before anything scans it, so a call loses the summary when redaction found a secret — a runner env value or a pattern's — in its arguments; pi, codex and OpenCode report a summary and no arguments, and there a value that straddles the cut stays in part. `execute_tool` span names and ids keep the pattern pipeline #6603 gave them, without the literal pass.
- **Out of scope:** pi/codex arguments and ids (#7414), raising the bounds (#7415), sub-agent nesting (`parent_tool_use_id` is still dropped at decode, so sub-agent calls and their arguments sit flat and unattributed on the record; it stays ADR 0050's deferred item 1, as ADR 0108's Consequences records — no issue filed yet), and the rest of the model's input: the fixed default prompt is a constant that carries no task and is left out by choice; the requests the runtime builds are not visible to fullsend, which reads the runtime's stream, not its API requests.
- **Related:** #7454 (EM-002, #7246) left its behavioural rules out because tool-call arguments were not on the trace. After this PR they are on the Level 3 message record only — content gate on, Claude runs, redacted, dropped whole over 8 KiB — and not on the always-on `execute_tool` spans EM-002 reads. Whether that is enough for those rules is for #7246 to decide.

## Evidence

Local gated run of this branch at `e343f872`, 2026-09-21. It predates the four review fixes below and was not repeated; up to `2a4ed4b9` only comments, docs and one test comment had changed. File sink only — no OTLP endpoint was set, so this shows the record, not backend acceptance (every attribute here is under 9 KB). A throwaway agent from `fullsend agent new --validation-loop`, with `feedback_mode: append` and a validator that rejects iteration 1 on purpose; `claude-opus-4-6` on Vertex; exit 0, validation passed on iteration 2. Trace `3e38910399f09ce77fcd10e5b6255655`, 26 spans. (An earlier run of the same recipe, before the last review fix, gave the same picture: trace `73d0767e85cce6ce24cd3547e93b3609`.)

| | iteration 1 `agent` span | iteration 2 `agent` span |
|---|---|---|
| `gen_ai.input.messages` | absent | 592 bytes: one `user` message, one text part — the default prompt, the framing, and the validator's two lines inside the `<validation-output>` fence |
| `gen_ai.output.messages` | 5,499 bytes, `finish_reason: stop` | 8,367 bytes, `stop` |
| `tool_call` parts carrying `arguments` | 10 of 10 | 12 of 12 |
| redactions, truncation | none | none |

A part as recorded: `{"type":"tool_call","id":"toolu_vrtx_01N3pu7GJbVVJbYMQaUFWPba","name":"Read","summary":"/sandbox/workspace/README.md","arguments":{"file_path":"/sandbox/workspace/README.md"}}`. The 22 `execute_tool` spans carry `gen_ai.operation.name`, `gen_ai.tool.name`, `gen_ai.tool.call.id` and, on six failed calls, `error.type` — nothing of the arguments. The gate-off case is covered by the pinned test, not by a live run.

## Tests

Red first, except where this paragraph says otherwise: three tests were green on arrival, and the `attachInput` call site has no unit test. 46 hand mutants over the new behaviour (charge, redaction, findings, role and part type, the `Result` guard, invalid-JSON handling, `UseNumber`, null, the bound and its boundary, charge unit, key redaction and single key scan, the per-member scan — removed, gated on a clean value, keeping any finding, uncounted, quotes kept, run through the whole pipeline — key collisions, sorted walk, nameless calls, array walk, accounting, single scan, serialized-scan regression, span leakage, both parser paths, renderer) — all killed. Pre-push review over three rounds found one high defect (the member-name scan was skipped when the value raised any other finding) and two regressions in its fixes (an underscore as the quote stand-in; the scanned pair going back through the normalizer), plus wording and accounting errors in the docs and comments — all fixed in these commits. Three tests were green on arrival, with no production code changed under them: arguments absent from `execute_tool` spans, and arguments and results absent from `run-telemetry.jsonl` with the gate off while the tool span is still written (both labelled as pins in their commit), and the renderer not printing arguments (`renderer.go` is untouched). The one line in `runAgent` that calls `attachInput` has no unit test, as with `attachContent`'s call site; the evidence run above exercises it. After opening, review led to four fixes, each red first. From Qodo: arguments lost to a redacted-away name were not charged or marked (`99f13991`), the delta crossing the parser's input cap went in whole (`1c124ceb`), and the collector had no literal pass for runner env values (`c697b84b`). From the review agent: that pass ran only ahead of the normalizer, so a value in compatibility forms reached the record (`a1c1851a`, which also drops the summary of a call whose arguments held a secret and scans number leaves). 25 more hand mutants over them, all killed. Two heavier designs for the last fix — a pipeline stage between the normalizer and the patterns, and masking a value's beginning at the end of a string — were built and dropped after review: the first let a combining mark after a value defeat it and its mask hid the keyword a pattern needs; the second masked ordinary words and dropped call ids.

### CODE REVIEW

File: internal/cli/content_collector.go [hunk 1]
```diff
@@ -622,6 +716,110 @@ func (c *contentCollector) redact(text string, findings *[]security.Finding) str
...
+// way, so each string member is scanned once more, already redacted,
+// beside its redacted key (secretNamed) and masked whole on a match.
+// Keys are walked in sorted order, so neither the findings nor a dropped
+// value's charge follow map order. *collided reports two keys of one
+// object that redact to the same string: keeping either member would
+// show a call the agent did not make.
+func (c *contentCollector) redactValue(v any, collided *bool) any {
+	switch t := v.(type) {
+	case string:
+		return c.redact(t, &c.findings)
```
[the remark this answers is not in the document]
Comment (dhshah13) [reply internal/cli/content_collector.go:809]: Fixed in 3d19a5fe. `redact` runs the literal pass first — `replaceEnvSecrets`, now shared with `redactFeedback` — so it covers text, results, arguments, names, summaries and the input. An id holding a value is dropped.

File: internal/cli/content_collector.go [hunk 2]
```diff
@@ -468,6 +557,11 @@ func (c *contentCollector) Result(finishReason string) contentResult {
 			p.Response = c.redact(p.Response, &res.Findings)
 		}
 		p.Name = c.redact(p.Name, &res.Findings)
+		if p.Name == "" {
+			// As in Handle, for a name that redacts to nothing. Arguments
+			// Handle had already dropped stay charged and marked.
+			p.Arguments = nil
```
[the remark this answers is not in the document]
Comment (dhshah13) [reply internal/cli/content_collector.go]: Fixed in d214da4e. Arguments lost to a name that redacts away are charged to `dropped_bytes`; the span and the part are marked.

File: internal/cli/content_collector.go [hunk 3]
```diff
@@ -309,7 +391,14 @@ func (c *contentCollector) Handle(evt agentruntime.AgentEvent) {
...
 	case agentruntime.ToolUseEvent:
-		c.appendPart(contentPart{Type: "tool_call", ID: boundedID(e.ID), Name: e.Name, Summary: e.Summary})
+		p := contentPart{Type: "tool_call", ID: boundedID(e.ID), Name: e.Name, Summary: e.Summary}
+		if e.Name != "" {
+			// The schema requires a name on a tool_call part. Arguments
+			// alone must not keep a nameless call, nor charge for one that
+			// appendPart then refuses.
+			p.Arguments, p.Truncated = c.toolArguments(e.Arguments)
+		}
+		c.appendPart(p)
```
[the remark this answers is not in the document]
Comment (dhshah13) [reply internal/cli/content_collector.go:420]: No change. `lint.yml` runs `go test -race -coverprofile=coverage.out ./...` on every PR; it is the `test` check here.

File: internal/runtime/claude_progress.go [hunk 4]
```diff
@@ -293,9 +293,10 @@ func parseClaudeStream(r io.Reader, onEvent func(AgentEvent)) error {
 			case "content_block_stop":
 				if currentToolName != "" {
 					onEvent(ToolUseEvent{
-						ID:      currentToolID,
-						Name:    currentToolName,
-						Summary: extractSafeContext(currentToolName, json.RawMessage(toolInputJSON.String())),
+						ID:        currentToolID,
+						Name:      currentToolName,
+						Summary:   extractSafeContext(currentToolName, json.RawMessage(toolInputJSON.String())),
+						Arguments: toolInputJSON.String(),
```
[the remark this answers is not in the document]
Comment (dhshah13) [reply internal/runtime/claude_progress.go:302]: Fixed in 5fb7fdff. The delta that crosses `maxToolInputSize` is cut at the cap.

File: internal/cli/content_collector.go [hunk 5]
```diff
@@ -622,6 +716,110 @@ func (c *contentCollector) redact(text string, findings *[]security.Finding) str
...
+		for i := range t {
+			t[i] = c.redactValue(t[i], collided)
+		}
+	case map[string]any:
+		out := make(map[string]any, len(t))
+		for _, k := range slices.Sorted(maps.Keys(t)) {
+			rk := c.redact(k, &c.findings)
+			e := c.redactValue(t[k], collided)
+			if s, ok := e.(string); ok && c.secretNamed(rk, s) {
+				e = "***"
```
[the remark this answers is not in the document]
Comment (dhshah13) [reply internal/cli/content_collector.go:826]: Intentional. The member scan reproduces `json_field`, which matches a quoted string of 8+ characters directly under the key — the same in results, text and feedback. Masking any value under such a key would take `max_tokens` and `author` with it; widening it belongs in `internal/security`. The dev guide says so ("It reaches string members only").

File: internal/cli/content_collector.go [hunk 6]
```diff
@@ -125,27 +147,74 @@ func attachContent(span trace.Span, res contentResult) {
...
+//
+// The prompt is cut upstream (maxFeedbackBytes), not here. Folding can
+// grow the copy about elevenfold, so its encoded size is charged to
+// maxEncoded: the two content attributes of one span together stay
+// within the size maxEncodedContentBytes is proven for.
+func (c *contentCollector) attachInput(span trace.Span, prompt string) {
+	if c == nil {
+		return
+	}
+	text := c.redact(prompt, &c.findings)
```
[the remark this answers is not in the document]
Comment (dhshah13) [reply internal/cli/content_collector.go:188]: Fixed in a1c1851a. `redact` repeats the literal pass on the pipeline's result, so the input and the output record are covered; a test feeds a fullwidth and a zero-width-split value through text, arguments and `attachInput`. Not covered, and pinned for the assignment case: a value written that way which a pattern also recognises is masked by that pattern. A call also loses its `summary` when redaction found a secret in its arguments — the parser cuts the summary before anything scans it.

File: internal/cli/content_collector.go [hunk 7]
```diff
@@ -611,6 +716,10 @@ func (c *contentCollector) redact(text string, findings *[]security.Finding) str
 	if text == "" {
 		return text
 	}
+	text, keys := replaceEnvSecrets(text, c.runnerEnv)
```
[the remark this answers is not in the document]
Comment (dhshah13) [reply internal/cli/content_collector.go]: Fixed in a1c1851a. The pass now also runs on the pipeline's result; tests cover the fullwidth and zero-width-split forms through text, arguments and `attachInput`.

File: docs/ADRs/0108-tool-call-span-topology.md [hunk 8]
```diff
@@ -121,3 +121,16 @@ convention bump.
   call ids and results ([#7414](https://github.com/fullsend-ai/fullsend/issues/7414)).
 - This settles the span-granularity question in #294; retention and access
   remain open there.
+
+## Annotations
+
+**2026-09-18 — Tool arguments on the message record:** the Decision's "full
+arguments next" is in place up to a per-call bound: `tool_call` parts on the
```
[the remark this answers is not in the document]
Comment (dhshah13) [reply docs/ADRs/0108-tool-call-span-topology.md:128]: Fixed in 7944734c. Both annotations are a short note and a link to the reference. No new ADR here: ADR 0108's Decision already puts arguments on the message record.

[the remark this answers is not in the document]
Comment (dhshah13) [reply docs/ADRs/0108-tool-call-span-topology.md:128]: Fixed in 7944734c. Both annotations are now a short note and a link.

File: docs/guides/infrastructure/distributed-tracing.md [hunk 9]
```diff
@@ -87,31 +90,57 @@ The variable name and accepted values follow the
...
+whole — a cut object is not JSON — and the marked `tool_call` part keeps
+its `id`, name and summary. The input message is not cut at this stage:
+the validation output inside it is cut at 10 KiB when the prompt is
+composed (the prompt then says `[truncated]`), and the two markers above
+do not describe it. The recorded copy is the redacted one, so it can
+differ from the prompt the agent received: compatibility characters are
+folded, which can make it about eleven times longer, and its encoded size
+counts toward the 255,000 bytes. A tool result whose stream line
 exceeds the parser's 1 MiB bound is kept as an empty, marked
 `tool_call_response` part — the call was answered, its content is lost and
```
[the remark this answers is not in the document]
Comment (dhshah13) [reply docs/guides/infrastructure/distributed-tracing.md:169]: Fixed in 7944734c. The consumer contract says a marked `tool_call` part lost its arguments whole, never in part. No second marker: part type and the absent `arguments` tell the two apart.

File: internal/cli/content_collector.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (dhshah13) [reply internal/cli/content_collector.go:809]: Follow-up: a1c1851a repeats the pass after the pipeline as well — its normalizer can spell out a value the first pass saw in compatibility forms. After a rebase, the fix above is c697b84b.

File: docs/guides/infrastructure/distributed-tracing.md [hunk 10]
```diff
@@ -87,31 +109,62 @@ The variable name and accepted values follow the
...
-255,000 bytes, because encoding adds 9–11% at these bounds on real runs
+bounds count raw bytes, except that arguments count as the
+re-encoded JSON described above, against their own bound and against the
+total. That encoding writes `<`, `>` and `&` as six bytes each, so
+arguments dense in those characters reach their bound at about a sixth of
+that size on the stream. The exported JSON string is bounded as well, at
+255,000 bytes, because encoding adds 8–10% at these bounds on real runs
 and up to six times on escape-dense content — a record over that is trimmed again,
 oldest first.
 Truncation is marked via `fullsend.content.truncated` on the span and
```
[the remark this answers is not in the document]
Comment (dhshah13) [reply docs/guides/infrastructure/distributed-tracing.md]: Added in 67d08abb: the reference now says, where the marker is introduced, that `fullsend.truncated` reads by part type — part of a tool result kept, none of a tool call's arguments.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
